### ISSUE TITLE
ignoreFiles and .stylelintignore not being respected

### ISSUE THREAD

**Issue by laytong** [stylelint/stylelint#1693]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A bug where ignoreFiles and .stylelintignore seem to be ignored.
I upgraded our project packages/dependencies and our stylelint postcss task seems to be including (and erroring) on files specified in the ignore. 
I have tried using a .stylelintignore instead but neither seem to be respected.

Our previous package versions before the upgrade are:
stylelint: `6.6.0` and style-lint-config: `9.0.0`

Our config is unchanged before/after the dependency update.

> Which rule, if any, is this issue related to?

Doesn't look to be related to a specific rule

> What stylelint configuration is needed to reproduce this issue?

```
{
  "extends": "stylelint-config-standard",
  "ignoreFiles": "./static/components/**",
  "rules": {
    "color-no-invalid-hex": true,
    "selector-no-id": true,
    "function-linear-gradient-no-nonstandard-direction": null,

    "rule-non-nested-empty-line-before": null,
    "comment-empty-line-before": null,
    "at-rule-empty-line-before": null,

    "selector-pseudo-element-colon-notation": "single",
    "string-quotes": "single",
    "function-url-quotes": "always",
    "font-family-name-quotes": "always-unless-keyword",

    "color-hex-length": null,
    "color-hex-case": null,

    "declaration-no-important": true,
    "selector-max-specificity": "0,4,3",
    "declaration-block-no-ignored-properties": null,
    "function-name-case": null,
    "no-extra-semicolons": true,
    "no-duplicate-selectors": true
  }
}
```

> Which version of stylelint are you using?

stylelint: `7.0.3`
stylelint-config-standard: `11.0.0`

> How are you running stylelint: 

via grunt-postcss plugin.
here is the config we use for that task
```module.exports = {
    options: {
        processors: [
            require('stylelint')(),
            require('postcss-reporter')({
                clearMessages: true,
                throwError: true,
            }),
            require('autoprefixer')({
                browsers: ['last 2 versions'],
            }),
        ],
    },
    dist: {
        src: 'static/css/apps.css',
    },
};```

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

A warning is flagged from a file within the ignored directory.

<img width="613" alt="screen shot 2016-07-21 at 08 59 42" src="https://cloud.githubusercontent.com/assets/9656959/17029788/81ea0540-4f21-11e6-8ed0-1948cd5ddd08.png">

**Comment by jeddy3** [comment]
The behaviour of `ignoreFiles` and `.stylelintignore` diverged in `7.0.0` to accommodate requests for `.stylelintignore` to behave like `.gitignore`.

If your using `.stylelintignore`, try:

```
static/components
```

Or, if your using `ignoreFiles` in your configuration object try:

```
./static/components/**/*.less
```

Hopefully that should work.

Further details can be found in the [docs](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md#ignorefiles).

**Comment by laytong** [comment]
Sadly this appears to make no difference. I am still getting the same error out using your recommendations.

**Comment by jeddy3** [comment]
Are you using both a `.stylelintignore` file *and* the `ignoreFiles` configuration object property?

If so, can you try one *without* the other.

**Comment by laytong** [comment]
I am not using both at the same time. I have tried both independently

**Comment by jeddy3** [comment]
Are you using a combined build and lint Grunt task?

If so, try [splitting them apart, or using postcss-import](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/postcss-plugin.md#usage-examples).

Ideally you should use [`grunt-stylelint`](https://github.com/wikimedia/grunt-stylelint)... just [waiting on](https://github.com/wikimedia/grunt-stylelint/pull/19) a `7.0.0` compatible release for that though.

**Comment by laytong** [comment]
Joining the task is not the issue because running the grunt task on it's own still produces the error.

I'll take a look at using grunt-stylelint when the compatible release is out.

**Comment by jeddy3** [comment]
> Joining the task is not the issue because running the grunt task on it's own still produces the error.

Are you using Grunt's globs to ignore the the components folder?

```
module.exports = { options: { processors: [ require('stylelint')(), require('postcss-reporter')({ clearMessages: true, throwError: true, }), dist: { src: 'static/**/.less!static/components/**/*.less', }, };
```

> I'll take a look at using grunt-stylelint when the compatible release is out.

Yep, I think that'll be best :)

**Comment by davidtheclark** [comment]
@laytong Please try using stylelint's CLI directly and see what happens.

**Comment by jacobmischka** [comment]
I seem to be having this problem as well.

```
$ cat .stylelintrc.json 
{
	"extends": "stylelint-config-standard",
	"ignoreFiles": [
		"./css/main.scss"
	],
	"rules": {
		"indentation": "tab",
		"color-hex-case": "lower",
		"at-rule-empty-line-before": ["always", {
			"except": ["blockless-group"]
		}],
		"rule-nested-empty-line-before": "always",
		"max-empty-lines": 3
	}
}

$ stylelint css/*

css/main.scss
 1:1  ✖  Unknown word   CssSyntaxError
```

**Comment by davidtheclark** [comment]
Please tell us about the directory structure involved: where the config is, where the files are, where `process.cwd()` is. All this stuff works in relation to each other, so we need to know about each part to debug. And the simpler the test case the better!

**Comment by jacobmischka** [comment]
```
[mischka@mischka-laptop test-stylelint]$ tree -a
.
├── css
│   └── main.scss
├── node_modules
│   └── stylelint-config-standard
│       ├── CHANGELOG.md
│       ├── index.js
│       ├── LICENSE
│       ├── package.json
│       └── README.md
├── package.json
└── .stylelintrc.json

3 directories, 8 files
[mischka@mischka-laptop test-stylelint]$ cat .stylelintrc.json 
{
	"extends": "stylelint-config-standard",
	"ignoreFiles": [
		"./css/main.scss"
	],
	"rules": {
		"indentation": "tab",
		"color-hex-case": "lower",
		"at-rule-empty-line-before": "always",
		"rule-nested-empty-line-before": "always",
		"max-empty-lines": 3
	}
}
[mischka@mischka-laptop test-stylelint]$ stylelint css/*

css/main.scss
 1:1  ✖  Unknown word   CssSyntaxError
```

**Comment by jacobmischka** [comment]
For the record, I tried with `.stylelintignore` and using the `--ignore-path` argument as well, none seemed to work.

**Comment by davidtheclark** [comment]
@jacobmischka I think your issue might be different than the one above. I'm going to open another issue for it and mention you.

**Comment by jacobmischka** [comment]
Ah, I feared the problem might be that ignored files are still being parsed. Thanks for helping get to the bottom of it.

Unfortunately, the syntax that's breaking postcss I don't have much control over, it's required for jekyll's sass compiling. I guess I could compile my sass using a separate build step instead, but I'd rather not.

Thanks again!

**Comment by davidtheclark** [comment]
@jacobmischka can you just change your input file glob?

**Comment by jacobmischka** [comment]
Well, I'm using an editor plugin that lints the file that's currently open, so not really without adding file ignoring functionality to that, as far as I'm aware. I don't really see much point in doing that when it should probably be ignored directly in stylelint anyway, right?

**Comment by davidtheclark** [comment]
> I don't really see much point in doing that when it should probably be ignored directly in stylelint anyway, right?

Yes, I was just suggesting a stopgap measure until the bug is fixed.

**Comment by jacobmischka** [comment]
Ah, yeah it's not really a huge deal until it's fixed, I can deal with the extra error. Thanks though.

**Comment by m-allanson** [comment]
@laytong I'm unable to replicate this issue. Are you still running into this problem? 

If you are, a slimmed down repo that demonstrates the problem would allow us to investigate further. 

Here's a quick example repo you could use as a starter: https://github.com/m-allanson/stylelint-issue-1693

**Comment by jaydenseric** [comment]
I have configured `.stylintignore` to ignore `/bundle`, and this works when running `stylelint '**/*.css'` from the project root.

For some reason, opening `/bundle/bundle.css` in Atom with [steelbrain/linter](https://github.com/steelbrain/linter) and [AtomLinter/linter-stylelint](https://github.com/AtomLinter/linter-stylelint) installed it lints the file. Is it because of the cwd issue, perhaps the plugins are running from outside of the project or something?

[AtomLinter/linter-eslint](https://github.com/AtomLinter/linter-eslint) works as expected with ESLint and `.eslintignore` ignoring `/bundle`.

**Comment by davidtheclark** [comment]
I'm going to close this for now. The issue has become confused with side issues and I think we still don't have very exact clarity on what's going on. If anyone would like to pursue the issue further, please create a more focused issue *and some failing test cases* to illustrate that the issue is not specific to the setup of your project.

**Comment by edwardfhsiao** [comment]
im using webpack with it, same problem occured.
my config looks like this
```js
  postcss: function(webpack) {
    return [
      stylelint({
        ignorePath: '../../node_modules/',
        extends: 'stylelint-config-standard',
        rules: stylelintRules,
      }),
      postcssImport({addDependencyTo: webpack}),
      cssnext({autoprefixer: {browsers: "ie >= 9, ..."}}),
      postcssReporter({clearMessages: true})
    ]
  }
```
it just wont ignore no matter what path I specifed

**Comment by edwardfhsiao** [comment]
ok, my temporary solution is separate node_modules and my assets(working directory)
```js
    {
      test: /\.css$/,
      exclude: /assets/,
      loader: ExtractTextPlugin.extract("style-loader", "css-loader")
    },
    {
      test: /\.css$/,
      exclude: /node_modules/,
      loader: ExtractTextPlugin.extract("style", "css!postcss")
    },
```

**Comment by vijayhardaha** [comment]
Do we have any solution to this? I am using the Stylelint VS code extension, my `.stylelintignore` is to ignore the folders that I want to be ignored. I am using extension version `1.2.3`
